File: ios/Mango/Mango/MobileEmbeddingProvider.swift
import Foundation
import onnxruntime_objc

// MARK: - MobileEmbeddingProvider

/// On-device embedding provider for iOS using ONNX Runtime with CoreML execution provider.
///
/// Runs all-MiniLM-L6-v2 INT8 quantized (ARM64) via the Apple Neural Engine / CPU.
/// Conforms to the UniFFI-generated EmbeddingProvider protocol.
///
/// Model I/O:
///   Inputs:  input_ids [1, 256], attention_mask [1, 256], token_type_ids [1, 256] (int64)
///   Output:  last_hidden_state [1, 256, 384] (float32) — requires mean pooling + L2 norm
///
/// The output tensor name is last_hidden_state (shape [batch, seq_len, 384]).
/// Mean pooling is applied masked by attention_mask, then L2 normalization produces
/// a unit-norm 384-dimensional embedding suitable for cosine similarity search.
class MobileEmbeddingProvider: EmbeddingProvider {
    private static let embeddingDim = 384
    private static let maxLength = 256

    private let session: ORTSession
    private let tokenizer: WordPieceTokenizer

    /// Initialize the ONNX Runtime session with CoreML EP and load the tokenizer.
    ///
    /// Loads model_optimized.onnx and tokenizer.json from the app bundle.
    /// Throws EmbeddingError if either resource is missing or fails to load.
    init() throws {
        // Locate bundled model and tokenizer resources
        guard let modelURL = Bundle.main.url(forResource: "model_optimized", withExtension: "onnx") else {
            throw EmbeddingError.modelNotFound
        }
        guard let tokenizerURL = Bundle.main.url(forResource: "tokenizer", withExtension: "json") else {
            throw EmbeddingError.tokenizerNotFound
        }

        // Initialize tokenizer from bundled tokenizer.json
        self.tokenizer = try WordPieceTokenizer(tokenizerURL: tokenizerURL)

        // Create ORT environment with warning-level logging
        let env = try ORTEnv(loggingLevel: .warning)

        // Configure session options with CoreML execution provider
        // CPUAndNeuralEngine: routes compute to Apple Neural Engine when available,
        // falls back to CPU for unsupported ops. MLProgram format is required for ANE.
        let options = try ORTSessionOptions()
        try options.appendCoreMLExecutionProvider(withOptionsV2: [
            "MLComputeUnits": "CPUAndNeuralEngine",
            "ModelFormat": "MLProgram"
        ])

        // Create the ONNX Runtime session
        self.session = try ORTSession(env: env, modelPath: modelURL.path, sessionOptions: options)

        print("[MobileEmbeddingProvider] Loaded ONNX model with CoreML EP")
    }

    // MARK: - EmbeddingProvider Conformance

    /// Embed a batch of texts. Returns a flat [Float] of length texts.count * 384.
    ///
    /// Called synchronously from the Rust actor thread via UniFFI (Rust wraps in spawn_blocking).
    /// On any per-text error, returns 384 zero floats for that text — does not crash or throw.
    func embed(texts: [String]) -> [Float] {
        var result = [Float]()
        result.reserveCapacity(texts.count * Self.embeddingDim)

        for text in texts {
            let embedding = embedSingle(text)
            result.append(contentsOf: embedding)
        }

        return result
    }

    // MARK: - Private Inference

    /// Run inference for a single text. Returns 384-dim L2-normalized embedding,
    /// or 384 zero floats on any error.
    private func embedSingle(_ text: String) -> [Float] {
        let zeros = [Float](repeating: 0.0, count: Self.embeddingDim)

        do {
            let tokens = tokenizer.tokenize(text, maxLength: Self.maxLength)

            // Create ORTValue tensors from Int64 arrays
            let inputIdsTensor = try makeTensor(tokens.inputIds)
            let attentionMaskTensor = try makeTensor(tokens.attentionMask)
            let tokenTypeIdsTensor = try makeTensor(tokens.tokenTypeIds)

            // Run ONNX Runtime inference
            let inputs: [String: ORTValue] = [
                "input_ids": inputIdsTensor,
                "attention_mask": attentionMaskTensor,
                "token_type_ids": tokenTypeIdsTensor
            ]
            let outputNames: Set<String> = ["last_hidden_state"]
            let outputs = try session.run(
                withInputs: inputs,
                outputNames: outputNames,
                runOptions: nil
            )

            guard let outputTensor = outputs["last_hidden_state"] else {
                return zeros
            }

            // Apply mean pooling masked by attention_mask, then L2 normalize
            return try meanPoolAndNormalize(outputTensor, attentionMask: tokens.attentionMask)
        } catch {
            print("[MobileEmbeddingProvider] embed error: \(error)")
            return zeros
        }
    }

    /// Create an ORTValue int64 tensor from an [Int64] array with shape [1, array.count].
    private func makeTensor(_ values: [Int64]) throws -> ORTValue {
        var mutableValues = values
        let data = NSMutableData(
            bytes: &mutableValues,
            length: mutableValues.count * MemoryLayout<Int64>.size
        )
        return try ORTValue(
            tensorData: data,
            elementType: .int64,
            shape: [1, NSNumber(value: values.count)]
        )
    }

    // MARK: - Mean Pooling + L2 Normalization

    /// Extract last_hidden_state [1, seqLen, 384], apply masked mean pooling,
    /// then L2-normalize to produce a unit-norm 384-dim embedding.
    private func meanPoolAndNormalize(_ outputTensor: ORTValue, attentionMask: [Int64]) throws -> [Float] {
        let embeddingDim = Self.embeddingDim
        let seqLen = Self.maxLength

        // Extract raw float32 data from [1, seqLen, 384] output tensor
        let rawData = try outputTensor.tensorData() as Data
        let floats = rawData.withUnsafeBytes { Array($0.bindMemory(to: Float.self)) }

        // Mean pool: sum token embeddings where attention_mask == 1, divide by count
        var mean = [Float](repeating: 0.0, count: embeddingDim)
        var count: Float = 0.0

        for i in 0..<seqLen {
            guard i < attentionMask.count, attentionMask[i] == 1 else { continue }
            let offset = i * embeddingDim
            guard offset + embeddingDim <= floats.count else { break }
            for d in 0..<embeddingDim {
                mean[d] += floats[offset + d]
            }
            count += 1.0
        }

        // Divide by attended token count (guard against zero-length input)
        let divisor = max(count, 1e-9)
        for d in 0..<embeddingDim {
            mean[d] /= divisor
        }

        // L2 normalize to unit norm (required for cosine similarity search)
        var norm: Float = 0.0
        for d in 0..<embeddingDim {
            norm += mean[d] * mean[d]
        }
        norm = sqrt(norm)
        let normDivisor = max(norm, 1e-12)
        for d in 0..<embeddingDim {
            mean[d] /= normDivisor
        }

        return mean
    }
}

File: android/app/src/main/java/dev/disobey/mango/MobileEmbeddingProvider.kt
package dev.disobey.mango

import android.content.Context
import ai.onnxruntime.OnnxTensor
import ai.onnxruntime.OrtEnvironment
import ai.onnxruntime.OrtSession
import dev.disobey.mango.rust.EmbeddingProvider
import java.io.File
import java.nio.LongBuffer
import kotlin.math.max
import kotlin.math.sqrt

/**
 * On-device embedding provider using ONNX Runtime with XNNPACK execution provider.
 *
 * Runs all-MiniLM-L6-v2 INT8 quantized model to produce L2-normalized 384-dimensional
 * embeddings for use in the RAG pipeline. Implements the UniFFI EmbeddingProvider
 * callback interface so the Rust core can dispatch embed() calls into native Kotlin.
 *
 * Phase 11, EMBD-03/EMBD-04/EMBD-05.
 */
class MobileEmbeddingProvider(context: Context) : EmbeddingProvider {
    private val ortEnv = OrtEnvironment.getEnvironment()
    private val session: OrtSession
    private val tokenizer: WordPieceTokenizer
    private val maxLength = 256
    private val embeddingDim = 384

    init {
        // ONNX Runtime requires a file path (not an InputStream) for session creation.
        // Copy model from assets to internal storage on first launch. Uses double-checked
        // locking so concurrent calls during app startup do not race on the file write.
        val modelFile = File(context.filesDir, "model_optimized.onnx")
        if (!modelFile.exists()) {
            synchronized(MobileEmbeddingProvider::class.java) {
                if (!modelFile.exists()) {
                    context.assets.open("model_optimized.onnx").use { input ->
                        modelFile.outputStream().use { output -> input.copyTo(output) }
                    }
                }
            }
        }

        // Build session options. Attempt to enable XNNPACK EP which accelerates
        // quantized INT8 ops on ARM CPUs. Falls back silently to default CPU EP
        // on devices where XNNPACK is unavailable.
        val options = OrtSession.SessionOptions()
        try {
            options.addXnnpack(
                mapOf("intra_op_num_threads" to Runtime.getRuntime().availableProcessors().toString())
            )
            android.util.Log.i("MobileEmbeddingProvider", "XNNPACK EP enabled")
        } catch (e: Exception) {
            android.util.Log.w(
                "MobileEmbeddingProvider",
                "XNNPACK EP unavailable, using CPU EP: ${e.message}"
            )
        }

        session = ortEnv.createSession(modelFile.absolutePath, options)

        // Load tokenizer vocabulary from bundled assets JSON
        val tokenizerJson = context.assets.open("tokenizer.json")
            .bufferedReader()
            .use { it.readText() }
        tokenizer = WordPieceTokenizer(tokenizerJson)

        android.util.Log.i(
            "MobileEmbeddingProvider",
            "Initialised: model=${modelFile.length() / 1_048_576}MB, vocab=${tokenizerJson.length / 1024}KB"
        )
    }

    /**
     * Embed a batch of texts. Returns a flat List<Float> of length texts.size * 384.
     * Each 384-float slice is an L2-normalised embedding for the corresponding text.
     * Returns 384 zero floats per text on per-text inference errors (does not crash).
     */
    override fun embed(texts: List<String>): List<Float> {
        val result = mutableListOf<Float>()
        for (text in texts) {
            try {
                val tokens = tokenizer.tokenize(text, maxLength)
                val embedding = runInference(tokens)
                for (v in embedding) result.add(v)
            } catch (e: Exception) {
                android.util.Log.e("MobileEmbeddingProvider", "Inference error for text: ${e.message}")
                // Fallback: return zero vector for this text so the pipeline does not crash
                repeat(embeddingDim) { result.add(0.0f) }
            }
        }
        return result
    }

    /**
     * Run a single-text ONNX inference pass and return a 384-float L2-normalised embedding.
     *
     * The model (all-MiniLM-L6-v2 ARM64 INT8) outputs last_hidden_state of shape
     * [1, seq_len, 384]. Mean pooling over unmasked token positions then L2 normalisation
     * produces a single embedding comparable to those from fastembed (DesktopEmbeddingProvider).
     */
    private fun runInference(tokens: TokenizerOutput): FloatArray {
        val shape = longArrayOf(1L, maxLength.toLong())

        val inputIdsTensor = OnnxTensor.createTensor(
            ortEnv, LongBuffer.wrap(tokens.inputIds), shape
        )
        val attentionMaskTensor = OnnxTensor.createTensor(
            ortEnv, LongBuffer.wrap(tokens.attentionMask), shape
        )
        val tokenTypeIdsTensor = OnnxTensor.createTensor(
            ortEnv, LongBuffer.wrap(tokens.tokenTypeIds), shape
        )

        val inputs = mapOf(
            "input_ids" to inputIdsTensor,
            "attention_mask" to attentionMaskTensor,
            "token_type_ids" to tokenTypeIdsTensor,
        )

        val results = session.run(inputs)

        // Release input tensors immediately after run to avoid GC pressure
        inputIdsTensor.close()
        attentionMaskTensor.close()
        tokenTypeIdsTensor.close()

        return try {
            // last_hidden_state: Array<Array<FloatArray>> shape [1][seq_len][384]
            @Suppress("UNCHECKED_CAST")
            val hiddenState = results[0].value as Array<Array<FloatArray>>
            val tokenEmbeddings = hiddenState[0]  // shape [seq_len][384]
            meanPoolAndNormalize(tokenEmbeddings, tokens.attentionMask)
        } finally {
            results.close()
        }
    }

    /**
     * Mean pool token embeddings over positions where attentionMask == 1L,
     * then L2-normalise the result. Matches fastembed mean pooling behaviour.
     */
    private fun meanPoolAndNormalize(
        tokenEmbeddings: Array<FloatArray>,
        attentionMask: LongArray
    ): FloatArray {
        val mean = FloatArray(embeddingDim)
        var count = 0f
        val seqLen = minOf(tokenEmbeddings.size, attentionMask.size)
        for (i in 0 until seqLen) {
            if (attentionMask[i] == 1L) {
                val tok = tokenEmbeddings[i]
                for (d in 0 until embeddingDim) {
                    mean[d] += tok[d]
                }
                count += 1f
            }
        }
        val divisor = max(count, 1e-9f)
        for (d in 0 until embeddingDim) {
            mean[d] /= divisor
        }
        return l2Normalize(mean)
    }

    /**
     * L2-normalise a vector in-place. Guards against division by near-zero norm
     * (epsilon 1e-12) to avoid NaN propagation when all token embeddings are zero.
     */
    private fun l2Normalize(vec: FloatArray): FloatArray {
        var norm = 0f
        for (v in vec) {
            norm += v * v
        }
        norm = sqrt(norm)
        val divisor = max(norm, 1e-12f)
        for (d in vec.indices) {
            vec[d] /= divisor
        }
        return vec
    }
}
